src/main/scala/firrtl/transforms/InferWidthsWithAnnos.scala

```diff
@@ -0,0 +1,95 @@
+// See LICENSE for license details.
+
+package firrtl
+package transforms
+
+import firrtl.annotations.{Annotation, ReferenceTarget, TargetToken}
+import firrtl.ir._
+import firrtl.Utils._
+import firrtl.passes.{InferWidths, MemPortUtils}
+
+case class WidthGeqConstraintAnnotation(loc: ReferenceTarget, exp: ReferenceTarget) extends Annotation {
+  def update(renameMap: RenameMap): Seq[WidthGeqConstraintAnnotation] = {
+    val newLoc :: newExp :: Nil = Seq(loc, exp).map { target =>
+      renameMap.get(target) match {
+        case None => Some(target)
+        case Some(Seq()) => None
+        case Some(Seq(one)) => Some(one)
+        case Some(many) =>
+          throw new Exception(s"Target below is an AggregateType, which " +
+            "is not supported by WidthGeqConstraintAnnotation\n" + target.prettyPrint())
+      }
+    }
+
+    (newLoc, newExp) match {
+      case (Some(l: ReferenceTarget), Some(e: ReferenceTarget)) => Seq(WidthGeqConstraintAnnotation(l, e))
+      case _ => Seq.empty
+    }
+  }
+}
+
+object InferWidthsWithAnnos extends Transform with ResolvedAnnotationPaths {
+  def inputForm: CircuitForm = UnknownForm
+  def outputForm: CircuitForm = UnknownForm
+
+  val annotationClasses = Seq(classOf[WidthGeqConstraintAnnotation])
+
+  def execute(state: CircuitState): CircuitState = {
+    val circuitName = state.circuit.main
+    val typeMap = new collection.mutable.HashMap[ReferenceTarget, Type]
+
+    def getComponentType(baseType: Type, tokens: Seq[TargetToken]): Type = {
+      if (tokens.isEmpty) {
+        baseType
+      } else {
+        val headType = tokens.head match {
+          case TargetToken.Index(idx) => sub_type(baseType)
+          case TargetToken.Field(field) => field_type(baseType, field)
+          case _: TargetToken.Ref => baseType
+        }
+        getComponentType(headType, tokens.tail)
+      }
+    }
+
+    def getDeclTypes(modName: String)(stmt: Statement): Unit = {
+      val pairOpt = stmt match {
+        case w: DefWire => Some(w.name -> w.tpe)
+        case r: DefRegister => Some(r.name -> r.tpe)
+        case n: DefNode => Some(n.name -> n.value.tpe)
+        case i: WDefInstance => Some(i.name -> i.tpe)
+        case m: DefMemory => Some(m.name -> MemPortUtils.memType(m))
+        case other => None
+      }
+      pairOpt.foreach { case (ref, tpe) =>
+        typeMap += (ReferenceTarget(circuitName, modName, Nil, ref, Nil) -> tpe)
+      }
```

Comment:
Why don't you update the typeMap directly in the stmt match?

Reply:
no real reason, `Some(ref -> tpe)` was just much less typing than `typeMap += (ReferenceTarget(circuitName, modName, Nil, ref, Nil) -> tpe)`

Reply:
It is a little bit DRYer this way